pymc_marketing/mmm/utility.py

```diff
@@ -0,0 +1,546 @@
+        - :math:`Q_{(1 - \\alpha)}` is the quantile at the specified confidence level.
+        - :math:`Q_{\\alpha}` is the quantile at the specified confidence level.
+
+    Parameters
+    ----------
+    confidence_level : float, optional
+        Confidence level for the quantiles (default is 0.75).
+
+    Returns
+    -------
+    UtilityFunction
+        A function that calculates the tail distance metric given samples and budgets.
+    """
+    if not 0 < confidence_level < 1:
+        raise ValueError("Confidence level must be between 0 and 1.")
+
+    def _tail_distance(
+        samples: pt.TensorVariable, budgets: pt.TensorVariable
+    ) -> pt.TensorVariable:
+        mean = pt.mean(samples)
+        q1 = _compute_quantile(samples, confidence_level)
+        q2 = _compute_quantile(samples, 1 - confidence_level)
+        return pt.abs(q1 - mean) + pt.abs(mean - q2)
+
+    return _tail_distance
+
+
+def _calculate_roas_distribution_for_allocation(
+    samples: pt.TensorVariable, budgets: pt.TensorVariable
+) -> pt.TensorVariable:
+    """Calculate the ROAS (Return on Advertising Spend) distribution for a given total budget.
+
+    This function computes the ratio of each sample (representing returns) to the sum of budgets.
+    The resulting distribution can be used to evaluate the efficiency of budget allocation across samples.
+
+    Parameters
+    ----------
+    samples : pt.TensorVariable
+        A 1D PyTensor tensor variable containing the returns for each asset or campaign.
+    budgets : pt.TensorVariable
+        A 1D PyTensor tensor variable representing the budget allocations for each asset or campaign.
+
+    Returns
+    -------
+    pt.TensorVariable
+        A PyTensor tensor variable representing the ROAS distribution.
+    """
+    total_budget = pt.sum(budgets)
+    roas_distribution = samples / total_budget
+    return roas_distribution
+
+
+def mean_tightness_score(
+    alpha: float = 0.5, confidence_level: float = 0.75
+) -> UtilityFunction:
+    R"""
+    Calculate the mean tightness score.
+
+    The mean tightness score is a risk metric that balances the mean return and the tail variability.
+    It is calculated as:
+
+    .. math::
+        Mean\ Tightness\ Score = \mu - \alpha \cdot Tail\ Distance
+
+    where:
+        - :math:`\mu` is the mean of the sample returns.
+        - :math:`Tail\ Distance` is the tail distance metric.
+        - :math:`\alpha` is the risk tolerance parameter.
+
+    alpha (Risk Tolerance Parameter): This parameter controls the trade-off.
+        - Higher :math:`\alpha` increases sensitivity to variability, making the metric value higher for spread dist
+        - Lower :math:`\alpha` decreases sensitivity to variability, making the metric value lower for spread dist
+
+    Parameters
+    ----------
+    alpha : float, optional
+        Risk tolerance parameter (default is 0.5).
+    confidence_level : float, optional
```

Comment:
We should add "confidence level must be between 0 and 1."

Reply:
Yes, good catch!